h/static/scripts/virtual-thread-list.js

```diff
@@ -0,0 +1,142 @@
+ *
+ * 'Virtualizing' the thread list improves UI performance by only creating
+ * annotation cards for annotations which are either in or near the viewport.
+ *
+ * Reducing the number of annotation cards that are actually created optimizes
+ * the initial population of the list, since annotation cards are big components
+ * that are expensive to create and consume a lot of memory. For Angular
+ * applications this also helps significantly with UI responsiveness by limiting
+ * the number of watchers (functions created by template expressions or
+ * '$scope.$watch' calls) that have to be run on every '$scope.$digest()' cycle.
+ *
+ * @param {Window} container - The Window displaying the list of annotation threads.
+ * @param {Thread} rootThread - The initial Thread object for the top-level
+ *        threads.
+ */
+function VirtualThreadList($scope, window_, rootThread) {
+  var self = this;
+
+  this._rootThread = rootThread;
+
+  // Cache of thread ID -> last-seen height
+  this._heights = {};
+
+  this.window = window_;
+
+  var debouncedUpdate = debounce(function () {
+    self._updateVisibleThreads();
+    $scope.$digest();
+  }, 20);
+  this.window.addEventListener('scroll', debouncedUpdate);
+  this.window.addEventListener('resize', debouncedUpdate);
+
+  this._detach = function () {
+    this.window.removeEventListener('scroll', debouncedUpdate);
+    this.window.removeEventListener('resize', debouncedUpdate);
+  };
+}
+inherits(VirtualThreadList, EventEmitter);
+
+/**
+ * Detach event listeners and clear any pending timeouts.
+ *
+ * This should be invoked when the UI view presenting the virtual thread list
+ * is torn down.
+ */
+VirtualThreadList.prototype.detach = function () {
+  this._detach();
+};
+
+/**
+ * Sets the root thread containing all conversations matching the current
+ * filters.
+ *
+ * This should be called with the current Thread object whenever the set of
+ * matching annotations changes.
+ */
+VirtualThreadList.prototype.setRootThread = function (thread) {
+  this._rootThread = thread;
+  this._updateVisibleThreads();
+};
+
+/**
+ * Sets the actual height for a thread.
+ *
+ * When calculating the amount of space required for offscreen threads,
+ * the actual or 'last-seen' height is used if known. Otherwise an estimate
+ * is used.
+ */
+VirtualThreadList.prototype.setThreadHeight = function (id, height) {
+  if (this._heights[id] === height) {
+    return;
+  }
+  this._heights[id] = height;
+};
+
+/**
+ * Recalculates the set of visible threads and estimates of the amount of space
+ * required for offscreen threads above and below the viewport.
+ *
+ * Emits a `changed` event with the recalculated set of visible threads.
+ */
+VirtualThreadList.prototype._updateVisibleThreads = function () {
+  // Space above the viewport in pixels which should be considered 'on-screen'
+  // when calculating the set of visible threads
+  var MARGIN_ABOVE = 800;
```

Comment:
I'd suggest moving these definitions to the top level of the file, so it's (even) clearer they're "constants".